xnmt/input_reader.py

```diff
@@ -68,11 +71,14 @@ def read_sent(self, line: str) -> xnmt.input.Input:
     raise RuntimeError("Input readers must implement the read_sent function")
 
   def count_sents(self, filename):
-    f = open(filename, encoding='utf-8')
-    try:
-      return sum(1 for _ in f)
-    finally:
-      f.close()
+    if filename in self.line_counts:
```

Comment:
There's a recent Python built-in annotator that does this and would save quite a few lines of code. One would simply put ```@functools.lru_cache``` before the method definition. See also 
https://docs.python.org/3.6/library/functools.html#functools.lru_cache